### PR TITLE
[LLD][COFF] Detect weak reference cycles.

### PR DESCRIPTION
**Pull request by cjacek**
Instead of entering an infinite loop.

### CODE REVIEW

**Comment by dpaoliello** [comment]
Is there any reason not to ICE when this happens?

File: lld/COFF/Symbols.cpp [hunk 1]
```diff
@@ -126,9 +127,14 @@ DefinedImportThunk::DefinedImportThunk(COFFLinkerContext &ctx, StringRef name,
 
 Defined *Undefined::getWeakAlias() {
   // A weak alias may be a weak alias to another symbol, so check recursively.
-  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias)
+  SmallSet<Symbol *, 4> weakChain;
+  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias) {
+    if (weakChain.contains(a))
+      break; // We have a cycle.
```
Comment (aganea) [inline lld/COFF/Symbols.cpp:133]: Instead of returning null then emitting "error: undefined symbol:" wouldn't it be nicer to say something like "can't resolve weak symbol because there's a cycle", and then "here's the cycle chain..." (with corresponding ref. obj files)?

**Comment by compnerd** [comment]
The weak aliases are an extension for MinGW right? In that case, what happens on ELF if you have weak alias cycles?

File: lld/test/COFF/weak-ref-cycle.test [hunk 2]
```diff
@@ -0,0 +1,42 @@
+REQUIRES: x86
```
Comment (MaskRay) [inline lld/test/COFF/weak-ref-cycle.test:1]: I think COFF doesn't have ELF's weak references. COFF only has weak externals, similar to ELF's weak definitions.

I'd use `weak-external-cycle.test` to be beside `weak-external{,2,3}.test`

File: lld/COFF/Symbols.cpp [hunk 3]
```diff
@@ -126,9 +127,14 @@ DefinedImportThunk::DefinedImportThunk(COFFLinkerContext &ctx, StringRef name,
 
 Defined *Undefined::getWeakAlias() {
   // A weak alias may be a weak alias to another symbol, so check recursively.
-  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias)
+  SmallSet<Symbol *, 4> weakChain;
```
Comment (MaskRay) [inline lld/COFF/Symbols.cpp:130]: DenseMap, which I believe has the least code size increase.

SmallPtrSet would be better than SmallSet but its small container case compiles to additional code than DenseMap.

File: lld/test/COFF/weak-ref-cycle.test [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+.set testsym, weak1
+    .weak weak1
+.set weak1, weak2
+    .section .test, "r"
+
+#--- weak2.s
+    .weak weak2
+.set weak2, weak1
+
+#--- weak-def.s
```
Comment (MaskRay) [inline lld/test/COFF/weak-ref-cycle.test:38]: I'd name this `def.s` since the definition is strong

File: lld/COFF/Symbols.cpp [hunk 5]
```diff
@@ -126,9 +127,14 @@ DefinedImportThunk::DefinedImportThunk(COFFLinkerContext &ctx, StringRef name,
...
+  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias) {
+    if (weakChain.contains(a))
+      break; // We have a cycle.
     if (auto *d = dyn_cast<Defined>(a))
       return d;
+    weakChain.insert(a);
```
Comment (grandinj) [inline lld/COFF/Symbols.cpp:136]: driveby comment: Normally it is faster to do something like

    if (!weakChain.insert(a).second)
        break; // we have a cycle

i.e. do the lookup&insert in one operation

**Comment by mstorsjo** [comment]
> The weak aliases are an extension for MinGW right? In that case, what happens on ELF if you have weak alias cycles?

No, weak aliases (formally, "weak externals") are a common MS COFF feature. Unlike ELF (where you either have a weak defined or a weak undefined symbol), the COFF weak externals are a pointer to another symbol, to substitute, if the original one ended up undefined.

But in MinGW environments, the COFF weak externals can be used to try to emulate ELF style weak symbols.

**Comment by compnerd** [comment]
Ah, okay, I didn't know that MinGW used weak externals to emulator weak symbols.

File: lld/COFF/Symbols.cpp (hunk 1 above)
Comment (cjacek) [reply lld/COFF/Symbols.cpp:133]: I tried to implement that, but there were a some problems with it. I pushed a draft here: https://github.com/cjacek/llvm-project/commit/55a76b765ff307c88a224201e088e50bb0aac89b

`getWeakAlias` callers don't expect errors from the call, so I had to use `fatal()` instead of `error()` to make it work like that. I guess it's what @dpaoliello meant by ICE, so maybe that's fine, I'm not sure.

We also don't track an input file associated with `Undefined` symbols, so we can't easily emit referencing files. We could potentially use something like `getSymbolLocations` (like we do in `reportUndefinedSymbol`), but we don't have access to the symbol table here without further refactoring. It seems to me that if we want something like that, it may be better to keep `getWeakAlias` returning null, detect in `reportUndefinedSymbol` that the reason for the symbol being undefined is a cycle and improve its reporting there. It would also avoid using `fatal()`.

FWIW, both MSVC linker and GNU ld report such symbols as undefined. link.exe does it pretty much like I did in this PR. GNU ld reports it only if the symbol is actually referenced, so I had to add something like `.rva testsym` to trigger it (that's not specific to cycles, other undefined weak externals are treated like that as well).

File: lld/test/COFF/weak-ref-cycle.test (hunk 2 above)
Comment (cjacek) [reply lld/test/COFF/weak-ref-cycle.test]: Renamed in the new version.

File: lld/test/COFF/weak-ref-cycle.test (hunk 4 above)
Comment (cjacek) [reply lld/test/COFF/weak-ref-cycle.test]: I renamed it. I also changed symbol names to sym1, sym2 to avoid a symbol with weak in its name being defined in this case.

File: lld/COFF/Symbols.cpp (hunk 5 above)
Comment (cjacek) [reply lld/COFF/Symbols.cpp]: Good point, I changed that in the new version (used `try_emplace` after changing set type to `DenseMap`).

File: lld/COFF/Symbols.cpp (hunk 3 above)
Comment (mstorsjo) [reply lld/COFF/Symbols.cpp]: So DenseMap is better here, than using DenseSet?

File: lld/COFF/Symbols.cpp [hunk 6]
```diff
@@ -13,6 +13,7 @@
 #include "lld/Common/Memory.h"
 #include "lld/Common/Strings.h"
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallSet.h"
```
Comment (mstorsjo) [inline lld/COFF/Symbols.cpp:16]: The include of `SmallSet.h` is leftover when using `DenseMap`. (And if we don't already do it, we probably should include `DenseMap.h` explicitly somewhere?)

**Review by mstorsjo — APPROVED** [review]
The change looks good overall to me, so LGTM. But let's wait another day if there are further comments on the choice of container etc.

File: lld/COFF/Symbols.cpp (hunk 6 above)
Comment (cjacek) [reply lld/COFF/Symbols.cpp:16]: I will remove it, we include both `DenseMap.h` and `DenseSet.h` in `InputFiles.h` (included from `Symbols.cpp`).

File: lld/COFF/Symbols.cpp (hunk 3 above)
Comment (cjacek) [reply lld/COFF/Symbols.cpp]: Looking at `DenseSet`, it's a thin wrapper around `DenseMap`, so it seems fine to me. I will change it, please let me know if I should revert.

**Comment by cjacek** [comment]
I pushed a new version with redundant include removed and using `DenseSet`. I also noticed that we could save one insert in cases where we find an alias by reordering checks, so I also changed that. Thanks for reviews!

**Review by MaskRay — APPROVED** [review]

**Review by mstorsjo — APPROVED** [review]
LGTM, this looks very tidy and straightforward now!
